### ISSUE TITLE
Generic support or unnecessary cast for Parcelables

### ISSUE THREAD

**Issue by novemberox** [frankiesardo/icepick#35]
Icepick for generic field that implements Parcelable generates sth like this

	  public Bundle restoreInstanceState(Object obj, Bundle state) {
	    ProductSwitcherActivity target = (ProductSwitcherActivity) obj;
	    if (state == null) {
	      return null;
	    }
	    Bundle savedInstanceState = state;

	    target.latestProduct = (T) savedInstanceState.getParcelable(BASE_KEY + "latestProduct");

	    return parent.restoreInstanceState(target, savedInstanceState);
	  }

where T is my generic/parcelable. However this cast it not need here. I briefly looked through source code and you might want remove this line: https://github.com/frankiesardo/icepick/blob/master/icepick-processor/src/main/java/icepick/processor/TypecastStrategy.java#L33

There is also an issue with `ArrayList` parceling, then class above adds two casts. While it might be needed for non parcel Objects it's not needed when retrieving `Parcelable` objects. In other words classes `TypecastStrategy` and `TypeToBundleMethodMap` probably should be tighter or `TypecastStrategy` should know what inspect what element is passsed?

**Comment by frankiesardo** [comment]
ArrayList double cast is necessary to force type erasure #25 

You're right, Parcelable cast is not needed, but it is part of the same typecast family and keeping it out is more a cognitive load than including it. Runtime penalty is basically non-existent.

If there are more pressing reasons to remove it I'm happy to give it a second though.

**Comment by novemberox** [comment]
Thanks for looking into it. Although you probably right about second part of my question, you haven't written what you think about generics support.

Let's say I have base `ProductActivity` class that is responsible to display `Product`. Now many classes inherits from `Product` so I want to declare my `Activity` as `ProductActivity<T extends Product>` and have `protected T product` variable that would be persisted using @Icepick. Having this situation Icepick would generate code like
 
    target.product= (T) savedInstanceState.getParcelable(BASE_KEY + "product");

The `T` type is unknown so this code won't compile. This cast is unnecessary, it might be impossible to have type mismatch here.

**Comment by frankiesardo** [comment]
You're right, that is actually a bug. I think the typecast should be smarter, otherwise simply removing "Parcelable" from the typecast list won't solve the issue. I'll keep this thread open

**Comment by IlyaEremin** [comment]
Any progress here?

**Comment by frankiesardo** [comment]
Yes, there is a new version ready to be published soon. Would you like to
try a snapshot version?

**Comment by IlyaEremin** [comment]
Yeap, I would be happy if you would upload snapshot.

**Comment by frankiesardo** [comment]
This is the release candidate:
```
dependencies {
  compile "frankiesardo:icepick:3.0.0-SNAPSHOT"
  provided "frankiesardo:icepick-processor:3.0.0-SNAPSHOT"
}
repositories {
    mavenCentral()
    maven {
      url "https://clojars.org/repo/"
    }
}
```

**Comment by frankiesardo** [comment]
Fixed with 460e7c82690977dea7e46be474567298b92a3da4
